raft/quorum/majority.go

```diff
@@ -0,0 +1,188 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quorum
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// MajorityConfig is a set of IDs that uses majority quorums to make decisions.
+type MajorityConfig map[uint64]struct{}
+
+// Describe returns a (multi-line) representation of the commit indexes for the
+// given lookuper.
+func (c MajorityConfig) Describe(l IndexLookuper) string {
+	if len(c) == 0 {
+		return "<empty majority quorum>"
+	}
+	type tup struct {
+		id, idx uint64
+		ok      bool // idx found?
+		bar     int  // length of bar displayed for this tup
+	}
+
+	// Below, populate .bar so that the i-th largest commit index has bar i (we
+	// plot this as sort of a progress bar). The actual code is a bit more
+	// complicated and also makes sure that equal index => equal bar.
+
+	n := len(c)
+	info := make([]tup, 0, n)
+	for id := range c {
+		idx, ok := l.Index(id)
+		info = append(info, tup{id: id, idx: idx, ok: ok})
+	}
+
+	// Sort by index
+	sort.Slice(info, func(i, j int) bool {
+		if info[i].idx == info[j].idx {
+			return info[i].id < info[j].id
+		}
+		return info[i].idx < info[j].idx
+	})
+
+	// Populate .bar.
+	for i := range info {
+		if i > 0 && info[i-1].idx < info[i].idx {
+			info[i].bar = i
+		}
+	}
+
+	// Sort by ID.
+	sort.Slice(info, func(i, j int) bool {
+		return info[i].id < info[j].id
+	})
+
+	var buf strings.Builder
+
+	// Print.
+	fmt.Fprint(&buf, strings.Repeat(" ", n)+"    idx\n")
+	for i := range info {
+		bar := info[i].bar
+		if !info[i].ok {
+			fmt.Fprint(&buf, "?"+strings.Repeat(" ", n))
+		} else {
+			fmt.Fprint(&buf, strings.Repeat("x", bar)+">"+strings.Repeat(" ", n-bar))
+		}
+		fmt.Fprintf(&buf, " %5d    (id=%d)\n", info[i].idx, info[i].id)
+	}
+	return buf.String()
+}
+
+// slicePool avoids allocations during commit index computations.
+var slicePool = sync.Pool{New: func() interface{} {
+	return make([]uint64, 0, 5)
+}}
+
+// CommittedIndex computes the committed index from those supplied via the
+// provided IndexLookuper. The outcome (a CommitRange) is final (meaning that
+// its two components agree) if enough voters are reflected in the
+// IndexLookuper. Otherwise, necessarily the value for Maybe is larger than that
+// for Defininitely (which is the largest index known committed based on the
+// information so far) and future values for Definitely may increase (limited by
+// Maybe) as previously missing voters are reflected in the IndexLookuper.
+func (c MajorityConfig) CommittedIndex(l IndexLookuper) CommitRange {
+	n := len(c)
+	if n == 0 {
+		return CommitRange{Definitely: math.MaxUint64, Maybe: math.MaxUint64}
+	}
+
+	srt := slicePool.Get().([]uint64)
```

Comment:
I'm pretty interested in whether any of these performance-related changes show up in `rafttoy`. If so, I don't think it would be too bad to cache the slice in the `MajorityConfig` struct and take it as a pointer receiver to this method. This would also allow us to avoid the reflection with `sort.Slice` without then requiring an allocation to box the slice through an interface.

Reply:
I was unable to find a performance regression in the numbers, but I did put in some extra work to actually remove the allocation here, PTAL at the updated first commit. 

I wrote the code to stash a slice on MajorityConfig but it ended up sort of nasty in that it cluttered the way in which one interacts with a MajorityConfig. Instead, I went ahead and implemented sorting on the stack. This won't allocate for <= 7 voters (as demonstrated by the benchmark, I'm happy to lower that to 5 if we want to save some stack space).